src/ZODB/FileStorage/FileStorage.py

```diff
@@ -672,28 +672,33 @@ def _data_find(self, tpos, oid, data):
         tid, tl, status, ul, dl, el = unpack(TRANS_HDR, h)
         status = as_text(status)
         self._file.read(ul + dl + el)
-        tend = tpos + tl + 8
+        tend = tpos + tl
```

Comment:
Here it looks to be correct - and this is another fix on it own: when data records of the transactions are iterated, the criteria is `while pos < tend` with pos initially pointing at first data record and advancing to the beginning of next data record on each iteration.

So let's see what happens on advance of the last data record:

1. pos = &data_record
2. pos += size(data_record)
3. check `pos < tend`

after "2" pos will point to the beginning of last `"- 8-byte redundant transaction length -8"` 

https://github.com/zopefoundation/ZODB/blob/6f8878124fa8bf1cc1932a8f67da8d15b653897c/src/ZODB/FileStorage/format.py#L53-L55

`tpos + tl + 8` points to _after_ that last `"- 8-byte redundant transaction length -8"`
`tpos + tl`  points to the _beginning_ of that last `"- 8-byte redundant transaction length -8"`

So if we check `pos < tpos + tl` we stop the loop after going through last data record.

But if we check `pos < tpos + tl + 8` after the last data record we will overrun the correct data records of the transaction and will use follow-up bytes treating them as another data record. We will be lucky if those bytes do not yield DataHeader with our oid, or else it might become a data corruption.

@perrinjerome am I right in understanding your rationale here?

Reply:
Thank you @navytux this was exactly this. If we stop at `pos < tpos + tl + 8` while iterating the data records, then we try to read the `- 8-byte redundant transaction length -8` and whatever is after that in the file as a data record, which is wrong.

That's true that this is also a fix on its own, I did not think about it this way. I guess that in normal conditions it's expected to find a data record for that oid, but if we don't find one, the behavior is incorrect.

Reply:
Jérome, thanks for feedback and for confirming my understanding is correct. Good to see we are on the common track here. Since this is another fix on its own maybe there is a way to add test for this particular fix as well? I do not insist on this, but if maybe that is easy, it makes sense to do. Thanks once again for bringing all this up.

Reply:
I tried to write a test calling restore with a txn that did not contain the oid, it triggers the problem, but it does not really represent a normal scenario so I'm not 100% sure what to do. While doing this I found another case when restore does not behave well, so I made another preparatory pull request https://github.com/zopefoundation/ZODB/pull/397 so that we can discuss only this ` + 8 ` story.